File: src/main/java/AoC2021_25.java
import static java.util.stream.Collectors.toSet;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2021_25 extends AoCBase {
    
    private static final char SOUTH = 'v';
    private static final char EAST = '>';
    private static final char EMPTY = '.';

    private final Set<Position> eastHerd;
    private final Set<Position> southHerd;
    private final int rows;
    private final int cols;
    
    private AoC2021_25(final List<String> input, final boolean debug) {
        super(debug);
        this.rows = input.size();
        this.cols = input.get(0).length();
        this.eastHerd = new HashSet<>();
        this.southHerd = new HashSet<>();
        IntStream.range(0, rows)
            .boxed()
            .flatMap(row -> IntStream.range(0, cols)
                                .mapToObj(col -> Position.of(row, col)))
            .forEach(p -> {
                final char ch = input.get(p.getX()).charAt(p.getY());
                if (ch == SOUTH) {
                    this.southHerd.add(p);
                } else if (ch == EAST) {
                    this.eastHerd.add(p);
                }
            });
    }
    
    public static final AoC2021_25 create(final List<String> input) {
        return new AoC2021_25(input, false);
    }

    public static final AoC2021_25 createDebug(final List<String> input) {
        return new AoC2021_25(input, true);
    }
    
    @SuppressWarnings("unused")
    private void print() {
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                final Position pos = Position.of(row, col);
                if (this.southHerd.contains(pos)) {
                    System.out.print(SOUTH);
                } else if (this.eastHerd.contains(pos)) {
                    System.out.print(EAST);
                } else {
                    System.out.print(EMPTY);
                }
            }
            System.out.println();
        }
        System.out.println();
    }
    
    private Position destinationEast(final Position p) {
        return Position.of(p.getX(), (p.getY() + 1) % cols);
    }
    
    private Position destinationSouth(final Position p) {
        return Position.of((p.getX() + 1) % rows, p.getY());
    }
    
    private Set<Position> findMovable(
            final Set<Position> herd,
            final Function<Position, Position> destination
        ) {
        return herd.stream()
            .filter(p -> {
                final Position newPos = destination.apply(p);
                return !this.eastHerd.contains(newPos) && !this.southHerd.contains(newPos);
            })
            .collect(toSet());
    }
    
    private void move(
            final Set<Position> toMove,
            final Set<Position> from,
            final Function<Position, Position> destination
    ) {
        toMove.stream()
            .forEach(p -> {
                from.remove(p);
                from.add(destination.apply(p));
            });
    }

    @Override
    public Integer solvePart1() {
        int cnt = 0;
        int moved = -1;
        while (moved != 0) {
            final Set<Position> toMoveEast = findMovable(this.eastHerd, this::destinationEast);
            move(toMoveEast, this.eastHerd, this::destinationEast);
            final Set<Position> toMoveSouth = findMovable(this.southHerd, this::destinationSouth);
            move(toMoveSouth, this.southHerd, this::destinationSouth);
            moved = toMoveEast.size() + toMoveSouth.size();
            cnt++;
        }
        return cnt;
    }

    @Override
    public Integer solvePart2() {
        return 0;
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2021_25.create(TEST).solvePart1() == 58;

        final Puzzle puzzle = Aocd.puzzle(2021, 25);
        puzzle.check(
            () -> lap("Part 1", () -> AoC2021_25.create(puzzle.getInputData()).solvePart1()),
            () -> lap("Part 2", () -> AoC2021_25.create(puzzle.getInputData()).solvePart2())
        );
    }

    private static final List<String> TEST = splitLines(
        "v...>>.vv>\r\n" +
        ".vv>>.vv..\r\n" +
        ">>.>v>...v\r\n" +
        ">>v>>.>.v.\r\n" +
        "v>v.vv.v..\r\n" +
        ">.>>..v...\r\n" +
        ".vv..>.>v.\r\n" +
        "v.v..>>v.v\r\n" +
        "....v..v.>"
    );
}

File: src/main/python/AoC2021_25.py
#! /usr/bin/env python3
#
# Advent of Code 2021 Day 25
#

from __future__ import annotations
import os
import time
from typing import Callable
from collections.abc import Generator
from aoc import my_aocd
import aocd


EMPTY = "."
EAST = ">"
SOUTH = "v"


class Grid:
    rows: int
    cols: int
    south_herd: set[tuple[int, int]]
    east_herd: set[tuple[int, int]]

    def __init__(self,
                 rows: int,
                 cols: int,
                 south_herd: set[tuple[int, int]],
                 east_herd: set[tuple[int, int]]
                 ) -> Grid:
        self.rows = rows
        self.cols = cols
        self.south_herd = south_herd
        self.east_herd = east_herd


def _parse(inputs: tuple[str]) -> Grid:
    rows = len(inputs)
    cols = len(inputs[0])
    south_herd = set[tuple[int, int]]()
    east_herd = set[tuple[int, int]]()
    for row, line in enumerate(inputs):
        for col, c in enumerate(line):
            if c == SOUTH:
                south_herd.add((row, col))
            elif c == EAST:
                east_herd.add((row, col))
            elif c != EMPTY:
                raise ValueError
    return Grid(rows, cols, south_herd, east_herd)


def _destination_south(grid: Grid, row: int, col: int) -> tuple[int, int]:
    return ((row + 1) % grid.rows, col)


def _destination_east(grid: Grid, row: int, col: int) -> tuple[int, int]:
    return (row, (col + 1) % grid.cols)


def _find_south(grid: Grid) -> Generator[tuple[int, int]]:
    for row, col in grid.south_herd:
        destination = _destination_south(grid, row, col)
        if destination not in grid.south_herd \
           and destination not in grid.east_herd:
            yield (row, col)


def _find_east(grid: Grid) -> Generator[tuple[int, int]]:
    for row, col in grid.east_herd:
        destination = _destination_east(grid, row, col)
        if destination not in grid.south_herd \
           and destination not in grid.east_herd:
            yield (row, col)


def _move_south(grid: Grid, to_move: Generator[tuple[int, int]]) -> int:
    cnt = 0
    new_herd = {cell for cell in grid.south_herd}
    for cell in to_move:
        new_herd.remove(cell)
        new_herd.add(_destination_south(grid, *cell))
        cnt += 1
    grid.south_herd = new_herd
    return cnt


def _move_east(grid: Grid, to_move: Generator[tuple[int, int]]) -> int:
    cnt = 0
    new_herd = {cell for cell in grid.east_herd}
    for cell in to_move:
        new_herd.remove(cell)
        new_herd.add(_destination_east(grid, *cell))
        cnt += 1
    grid.east_herd = new_herd
    return cnt


def _print(grid: Grid) -> None:
    if not __debug__:
        return
    for row in range(grid.rows):
        for col in range(grid.cols):
            if (row, col) in grid.south_herd:
                print(SOUTH + ' ', end='')
            elif (row, col) in grid.east_herd:
                print(EAST + ' ', end='')
            else:
                print(EMPTY + ' ', end='')
        print()
    print()


def _solve(inputs: tuple[str], on_step: Callable) -> int:
    grid = _parse(inputs)
    cnt = 0
    moved = -1
    while moved:
        moved = _move_east(grid, _find_east(grid)) \
                + _move_south(grid, _find_south(grid))
        cnt += 1
        if on_step is not None:
            on_step(grid)
    return cnt


def part_1(inputs: tuple[str]) -> int:
    return _solve(inputs, None)


def part_2(inputs: tuple[str]) -> int:
    return None


def visualize_part_1(inputs: tuple[str]) -> None:
    step = 0

    def on_step(grid: Grid) -> None:
        nonlocal step
        step += 1
        time.sleep(0.25)
        os.system('cls' if os.name in ('nt', 'dos') else 'clear')  # nosec
        print(f"Step {step}:")
        _print(grid)

    print("Initial:")
    _solve(inputs, on_step)


TEST = """\
v...>>.vv>
.vv>>.vv..
>>.>v>...v
>>v>>.>.v.
v>v.vv.v..
>.>>..v...
.vv..>.>v.
v.v..>>v.v
....v..v.>
""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2021, 25)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 58
    assert visualize_part_1(TEST) is None

    inputs = my_aocd.get_input(puzzle.year, puzzle.day, 137)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()
